### PR TITLE
[Perf] Reduce SpaceCardsViewModel Creation

### PR DESCRIPTION
**Pull request by EdwardJTL**
One perf issue identified:
- a new `SpaceCardsViewModel` is created each time `SpaceCardsView` is re-rendered
- each `SpaceCardsView` is persistent
- each `SpaceCardsView` has a sink on `SpaceCardModel.allDetails` where it sorts all spaces
- as a result, `sortSpaces` takes up lots of CPU and memory access

See screenshots for before and after profiling. This should improve performance when app is running in foreground where card grid is updated often. I don't think this well help with the case where app has 20 sec+ load time lol.

Changes:
- Make `SpaceCardModel`, which we seem to only have one instance of, create and own a single `SpaceCardsViewModel`. the `SpaceCardsView` will just access that property and use it as it's `ObservedObject`

### How was this tested?
- [x] I tested this manually
- [ ] I added automated tests
- [ ] Existing automated tests are enough to cover my changes

### Manual test cases
Profiling flow
- Profile with CPU profiler and allocation list
- Launch app, wait til webpage finishes loading, mark generation
- Go to spaces in tab manager, mark generation
- Change a sort option (specifically, toggle ascending/descending for last updated), mark generation
- exit tab switcher, mark generation

### Screenshots and screen recordings
| Profiler Track | Before | After |
| ------- | ------- | ------- |
| CPU | Screen Shot 2022-07-12 at 11 47 09 AM | Screen Shot 2022-07-12 at 11 56 38 AM |
| Allocation | Screen Shot 2022-07-12 at 11 47 00 AM | Screen Shot 2022-07-12 at 11 56 53 AM |

## Issues addressed
* Link(s) to Github issue(s) (if applicable)

### CODE REVIEW

**Comment by EdwardJTL** [comment]
had to fix a bug with pinned spaces. 
when a space becomes pinned pinned, `indexInStore  != indexInDetails ` check in `listenManagerState` does not pass, and the array struct is not updated, so no filtering is triggered. @jonathangarelick i'm not sure if this is a bug

As a result, I added a manual trigger to fire another around of filtering to pick up the effects of `pinned` ordering.

**Comment by EdwardJTL** [comment]
also, there was a reference cycle issue in `SpaceCardsViewModel`. Removing the reference cycle is enough to resolve the issue; the objects will keep getting created but only the latest one will persist, so there's only one that's actually running the filtering code. it will run the filter each time a new model is created. 
However, that seems more like exploiting our poor rendering cycle than actually fixing the bug.

**Comment by yozuysal** [comment]
thanks for digging into this Edward. Should we consider

1) reverting the sorting change as a whole
2) addressing the single instance as you have here

then looking into how we can revert the revert for the sorting PR?

Since we are not 100% happy with the current approach as well, that feature is hidden enough and wouldn't worth risking designing as we address perf issues.

**Comment by yozuysal** [comment]
actually I take that back. what if we use SpaceStore as an observable object here so that we react to SpaceStore.State changes?

File: Client/Frontend/Browser/Card/CardModels.swift [hunk 1]
```diff
@@ -549,6 +549,8 @@ class SpaceCardModel: CardModel {
     @Published var sortType: SpaceSortState = .updatedDate
     @Published var sortOrder: SpaceSortOrder = .descending
 
+    private(set) var cardsViewModel: SpaceCardsViewModel!
```
[the remark this answers is not in the document]
Comment (EdwardJTL) [reply Client/Frontend/Browser/Card/CardModels.swift:552]: it's initialized in init so it should be fine...

**Comment by EdwardJTL** [comment]
> actually I take that back. what if we use SpaceStore as an observable object here so that we react to SpaceStore.State changes?

we could have `SpaceCardsViewModel` observe that and use `SpaceCardModel.allDetails` as a data source. but there might still be synchronization issues between when the `state` is set and when the new data is populated in `SpaceCardModel.allDetails`.

if we want to address the perf issue asap, we can just revert the sorting change and then take our time to re-design the architecture for the data flow.

**Comment by yozuysal** [comment]
lets prioritize addressing the perf change yeah. I am OK with reverting for now and we can revert revert later

**Pull request closed without merging**